File: backend/src/application/dtos/dtos.py
import uuid
from datetime import date, datetime
from enum import Enum

from pydantic import BaseModel, Field


class DTO(BaseModel):
    pass


class MoneyDTO(DTO):
    amount: int
    currency: str


class BudgetStrategyDTO(DTO):
    type: str  # e.g., 'monthly' or 'yearly'
    parameters: dict = Field(default_factory=dict)


class CategoryDTO(DTO):
    id: uuid.UUID
    budget_id: uuid.UUID
    name: str
    limit: MoneyDTO


# DTO for budget aggregate.
class BudgetDTO(DTO):
    id: uuid.UUID
    user_id: uuid.UUID
    total_limit: MoneyDTO
    currency: str
    start_date: date
    end_date: date
    strategy: BudgetStrategyDTO
    name: str
    deactivation_date: datetime | None = None
    categories: list[CategoryDTO] = []


# Enum for transaction types.
class TransactionTypeEnum(str, Enum):
    EXPENSE = "EXPENSE"
    INCOME = "INCOME"


# DTO for transaction aggregate.
class TransactionDTO(DTO):
    id: uuid.UUID
    category_id: uuid.UUID
    amount: MoneyDTO
    transaction_type: TransactionTypeEnum
    occurred_date: datetime
    description: str | None = None
    user_id: uuid.UUID


# DTO for statistics record for a category.
class CategoryStatisticsRecordDTO(DTO):
    id: uuid.UUID
    category_id: uuid.UUID
    current_balance: MoneyDTO
    daily_available_amount: MoneyDTO
    daily_average: MoneyDTO
    used_limit: MoneyDTO


# DTO for overall statistics record.
class StatisticsRecordDTO(DTO):
    id: uuid.UUID
    user_id: uuid.UUID
    budget_id: uuid.UUID
    current_balance: MoneyDTO
    daily_available_amount: MoneyDTO
    daily_average: MoneyDTO
    used_limit: MoneyDTO
    creation_date: datetime
    categories_statistics: list[CategoryStatisticsRecordDTO] = Field(
        default_factory=list
    )

File: frontend/types/dtos.ts
import type { Money } from './money'

/**
 * DTO types from API using snake_case naming convention
 */

// DTO types related to Budget
export interface BudgetDTO {
  id: string
  user_id: string
  name: string
  total_limit: Money
  currency: string
  start_date: string
  end_date: string
  strategy: BudgetStrategyDTO
  deactivation_date: string | null
  categories: CategoryDTO[]
}

export interface BudgetStrategyDTO {
  type: 'monthly' | 'yearly' | string
  recurring: boolean
}

export interface PaginatedItems<T> {
  items: T[]
  total: number
  skip: number
  limit: number
}

// DTO types related to Transaction
export interface TransactionDTO {
  id: string
  category_id: string
  user_id: string
  amount: Money
  type: 'INCOME' | 'EXPENSE'
  date: string
  description?: string | null
}

// DTO types related to Category
export interface CategoryDTO {
  id: string
  name: string
  limit: Money
}

// DTO types related to API Payloads
export interface AmountPayload {
  amount: number
}

export interface CreateTransactionPayload {
  category_id: string
  amount: AmountPayload
  transaction_type: 'INCOME' | 'EXPENSE'
  occurred_date: Date
  description?: string
}

export interface UpdateTransactionPayload {
  category_id: string
  amount: AmountPayload
  transaction_type: 'INCOME' | 'EXPENSE'
  occurred_date: string
  description?: string
}

export interface MoneyPayload {
  amount: number
  currency: string
}

export interface StrategyPayload {
  budget_strategy_type: 'monthly' | 'yearly'
  parameters: Record<string, unknown>
}

export interface CreateCategoryRequestPayload {
  name: string
  limit: MoneyPayload
}

export interface CreateBudgetRequestPayload {
  name: string
  total_limit: MoneyPayload
  start_date: Date
  categories: CreateCategoryRequestPayload[]
  strategy: StrategyPayload
}

// DTO types related to Statistics
export interface CategoryStatisticsDTO {
  category_id: string
  current_balance: Money
  daily_available_amount: Money
  daily_average: Money
  used_limit: Money
}

export interface BudgetStatisticsDTO {
  current_balance: Money
  daily_available_amount: Money
  daily_average: Money
  used_limit: Money
  categories_statistics: CategoryStatisticsDTO[]
}

export interface DomainError {
    status: string; // Domain error type
    message: string; // User-friendly error message
  }
